File: src/longest_palindrome_by_concatenating_two_letter_words_2131/solution.go
package longest_palindrome_by_concatenating_two_letter_words_2131

import "fmt"

func longestPalindrome(words []string) int {
	wordToCount := make(map[string]int)

	for _, word := range words {
		wordToCount[word]++
	}

	putOneInCenter := false
	result := 0
	visited := make(map[string]struct{})

	for word, count := range wordToCount {
		if _, found := visited[word]; found {
			continue
		}

		visited[word] = struct{}{}

		if word[0] == word[1] {
			if count%2 == 0 {
				result += count * 2
			} else {
				result += (count - 1) * 2
				if !putOneInCenter {
					putOneInCenter = true
					result += 2
				}
			}
		} else {
			pair := string(word[1]) + string(word[0])
			if pairCount, found := wordToCount[pair]; found {
				visited[pair] = struct{}{}

				result += 2 * 2 * min(count, pairCount)
			}
		}
	}

	return result
}

func Test() {
	fmt.Println(longestPalindrome([]string{"lc", "cl", "gg"}))
	fmt.Println(longestPalindrome([]string{"ab", "ty", "yt", "lc", "cl", "ab"}))
	fmt.Println(longestPalindrome([]string{"cc", "ll", "xx"}))
}

File: src/longest_palindrome_by_concatenating_two_letter_words_2131/solution.ts
function longestPalindrome(words: string[]): number {
  const wordToCount = new Map<string, number>();

  for (const word of words) {
    wordToCount.set(word, (wordToCount.get(word) || 0) + 1);
  }

  const visited = new Set<string>();
  let usedCenter = false;
  let result = 0;

  for (const [word, count] of wordToCount.entries()) {
    if (visited.has(word)) {
      continue;
    }

    visited.add(word);

    if (word[0] === word[1]) {
      if (count % 2 === 0) {
        result += count * 2;
      } else {
        result += (count - 1) * 2;
        if (!usedCenter) {
          usedCenter = true;
          result += 2;
        }
      }
    } else {
      const pair = word[1] + word[0];
      const pairCount = wordToCount.get(pair) || 0;

      if (pairCount > 0) {
        visited.add(pair);
        result += 2 * 2 * Math.min(count, pairCount);
      }
    }
  }

  return result;
}

export function test() {
  console.log(longestPalindrome(["lc", "cl", "gg"]));
  console.log(longestPalindrome(["ab", "ty", "yt", "lc", "cl", "ab"]));
  console.log(longestPalindrome(["cc", "ll", "xx"]));
}
